src/components/tools/table-tool/use-row-height.ts

```diff
@@ -0,0 +1,65 @@
+import { useCallback } from "react";
+import { RowHeightArgs } from "react-data-grid";
+import { kCellHorizontalPadding, kCellLineHeight, kCellVerticalPadding,
+  kHeaderCellHorizontalPadding, kHeaderRowHeight, kRowHeight, TRow } from "./table-types";
+import { useCurrent } from "../../../hooks/use-current";
+import { measureTextLines } from "../hooks/use-measure-text";
+import { defaultFont } from "../../constants";
+import { IAttribute } from "../../../models/data/attribute";
+import { IDataSet } from "../../../models/data/data-set";
+import { ToolTileModelType } from "../../../models/tools/tool-tile";
+import { TableContentModelType } from "../../../models/tools/table/table-content";
+
+interface IUseRowHeight {
+  dataSet: IDataSet;
+  measureColumnWidth: (attr: IAttribute) => number;
+  model: ToolTileModelType;
+}
+export const useRowHeight = ({ dataSet, measureColumnWidth, model }: IUseRowHeight) => {
+  const modelRef = useCurrent(model);
+  const getContent = useCallback(() => modelRef.current.content as TableContentModelType, [modelRef]);
+  const content = getContent();
+
+  const textHeight = (text: string, width: number, font = defaultFont) => {
+    if (text) {
+      const cellHeight = measureTextLines(text.toString(), width, font) * kCellLineHeight
+        + 2 * kCellVerticalPadding;
+      return cellHeight;
+    }
+    return kRowHeight;
+  };
+
+  const rowHeight = useCallback((args: RowHeightArgs<TRow>) => {
+    let height = kRowHeight;
+    if (args.row) {
+      for (const [attrId, text] of Object.entries(args.row)) {
+        if (attrId !== '__context__' && attrId !== '__id__' && attrId !== '__index__') {
+          height = Math.max(height, textHeight(
+            text as string,
+            measureColumnWidth(dataSet.attrFromID(attrId)) - kCellHorizontalPadding,
+            defaultFont
+          ));
+        }
+      }
+    }
+    return height;
+  }, [measureColumnWidth, dataSet]);
```

Comment:
When does RDG call this? Is it called every time the row or column data changes? Or perhaps every time a row is rendered?

I'd hope RDG is smart and only calls it on rows that are visible and have changed (using some metric of "changed").

Did you look for a way to trigger RDG to re-call this? I'm just wondering if something like a cell renderer could identify how much height the browser thinks it needs to render without a scroll bar. And then we save that value into the RDG cell object (a property that isn't serialized), then we trigger a call to rowHeight. And that call to rowHeight can then use these browser computed heights instead of having to have our own function which uses the text and has to estimate how it will wrap.